README.rst
=======================
 ftw.maintenanceserver
=======================

Provides a simple server serving a static directory, which can be configured
as haproxy backup server, kicking in when all normal servers are offline.


How it works
------------

It is simple HTTP server, serving a static directory.
All requests which match no file in the static directory are responded with
the ``index.html`` which is expected to be placed in the static directory.

This is useful when a user is visiting the page at any URL while the maintenance
server is activated. The user then stays at the same URL and can refresh until
the system is back online and he is at the same place as he was before.


Installation
------------

The server can be installed using zc.buildout, which generates a pre configured
``bin/maintenance`` script::

    [buildout]
    parts += maintenance

    [maintenance]
    recipe = zc.recipe.egg
    eggs = ftw.maintenanceserver
    arguments = '${buildout:directory}/static', 8088


HAProxy configuration
---------------------

When using HAProxy the server can be simply configured as ``backup`` server,
which will only be used when all "normal" servers are offline::

    backend plone
    server plone1 127.0.0.1:8080 cookie p1 check downinter 15s maxconn 5 rise 1 slowstart 60s
    server plone2 127.0.0.1:8081 cookie p2 check downinter 15s maxconn 5 rise 1 slowstart 60s
    server maintenance 127.0.0.1:8088 check backup



Links
=====

- Github: https://github.com/4teamwork/ftw.maintenanceserver
- Issues: https://github.com/4teamwork/ftw.maintenanceserver/issues
- Pypi: http://pypi.python.org/pypi/ftw.maintenanceserver
- Continuous integration: https://jenkins.4teamwork.ch/search?q=ftw.maintenanceserver


Copyright
=========

This package is copyright by `4teamwork <http://www.4teamwork.ch/>`_.

``ftw.maintenanceserver`` is licensed under GNU General Public License, version 2.


docs/HISTORY.txt
Changelog
=========


1.1.3 (unreleased)
------------------

- Nothing changed yet.


1.1.2 (2017-11-28)
------------------

- Respond with 200 OK to resources request. [njohner]
- Fix tests to be compatible with most recent ftw.testbrowser. [lgraf]


1.1.1 (2017-02-03)
------------------

- Avoid caching maintenance server responses. [jone]


1.1.0 (2016-03-12)
------------------

- Allow setting a different listening address as optional parameter. Default is still localhost. pass an empty string to listen to all addresses on a host.
  [fredvd]


1.0.2 (2015-01-21)
------------------

- Fix broken pipe error when connections are closed early by the client.
  [jone]


1.0.1 (2014-03-21)
------------------

- Ignore virtualHostMonster configuartions in paths.
  [jone]


1.0.0 (2014-03-21)
------------------

- Initial implementation.
  [jone]


ftw/maintenanceserver/tests/test_server.py
from ftw.maintenanceserver.testing import SERVER_LAYER
from ftw.testbrowser import browser
from ftw.testbrowser import browsing
from StringIO import StringIO
from unittest2 import TestCase
import os.path
import sys


def catch_stderr(func):
    def stderr_catcher(*args, **kwargs):
        ori_stderr = sys.stderr
        sys.stderr = StringIO()
        try:
            return func(*args, **kwargs)
        except:
            print sys.stderr.getvalue()
            raise
        finally:
            sys.stderr = ori_stderr
    return stderr_catcher


class TestServer(TestCase):
    layer = SERVER_LAYER

    def open(self, path, method='GET'):
        url = os.path.join(self.layer['URL'], path)
        browser.open(url, method=method)

    def setUp(self):
        # We actually *expect* the maintenance server to produce
        # "HTTP Errors", such as "503 Service Unavailable"
        browser.raise_http_errors = False

    @browsing
    @catch_stderr
    def test_server_serves_index_html(self, browser):
        self.open('')
        self.assertEquals('Maintenance', browser.css('h1').first.text)

    @browsing
    @catch_stderr
    def test_directory_requests_falls_back_to_index(self, browser):
        self.open('images')
        self.assertEquals('Maintenance', browser.css('h1').first.text)

    @browsing
    @catch_stderr
    def test_resources_are_served(self, browser):
        self.open('images/logo.png')
        self.assertEquals('TheLogo', browser.contents.strip())

    @browsing
    @catch_stderr
    def test_restricted_to_document_root(self, browser):
        self.open('../__init__.py')
        self.assertEquals('Maintenance', browser.css('h1').first.text)

    @browsing
    @catch_stderr
    def test_POST_request_falls_back_to_GET(self, browser):
        self.open('', method='POST')
        self.assertEquals('Maintenance', browser.css('h1').first.text)

    @browsing
    @catch_stderr
    def test_HEAD_request(self, browser):
        # Respond to HEAD with 503 so that it is not cached.
        self.open('', method='HEAD')
        self.assertEquals(503, browser.status_code)
        self.assertEquals('text/html',
                          browser.headers.get('Content-Type'))

    @browsing
    @catch_stderr
    def test_OPTIONS_request(self, browser):
        # Options must be answered with 200 OK, otherwise HaProxy would
        # take the backend offline.
        # But since we want Varnish not to cache it, no matter how it is
        # configured, we set a cache-control header.
        self.open('', method='OPTIONS')
        self.assertEquals(200, browser.status_code)
        self.assertEquals('GET,HEAD,POST,OPTIONS',
                          browser.headers.get('Allow'))
        self.assertEquals('no-cache',
                          browser.headers.get('Cache-Control'))

    @browsing
    @catch_stderr
    def test_GET_responded_with_503(self, browser):
        # Respond to GET with 503 so that it is not cached.
        self.open('', method='GET')
        self.assertEquals(503, browser.status_code)
        self.assertEquals('Service Unavailable', browser.status_reason)

    @browsing
    @catch_stderr
    def test_POST_responded_with_503(self, browser):
        # Respond to POST with 503 so that it is not cached.
        self.open('', method='POST')
        self.assertEquals(503, browser.status_code)
        self.assertEquals('Service Unavailable', browser.status_reason)

    @browsing
    @catch_stderr
    def test_virtualHostMonster_configuration_is_removed(self, browser):
        self.open('VirtualHostBase/http/localhost:8080/mountpoint/'
                  'Plone/VirtualHostRoot/images/logo.png')
        self.assertEquals('TheLogo', browser.contents.strip())

    @browsing
    @catch_stderr
    def test_virtualHostMonster_inside_out_urls(self, browser):
        self.open('VirtualHostBase/http/localhost:8080/mountpoint/'
                  'Plone/VirtualHostRoot/_vh_the/_vh_site/images/logo.png')
        self.assertEquals('TheLogo', browser.contents.strip())

    @browsing
    @catch_stderr
    def test_resources_get_200(self, browser):
        self.open('images/logo.png')
        self.assertEquals(200, browser.status_code)


ftw/maintenanceserver/testing.py
from ftw.maintenanceserver.server import HTTPServer
from ftw.testbrowser import REQUESTS_BROWSER_FIXTURE
from plone.testing import Layer
from threading import Thread
import os


DOCUMENT_ROOT = os.path.join(os.path.dirname(__file__), 'tests', 'htdocs')
ADDRESS = 'localhost'
PORT = int(os.environ.get('ZSERVER_PORT', 55001))


class ServerLayer(Layer):

    defaultBases = (REQUESTS_BROWSER_FIXTURE, )

    def setUp(self):
        self.httpd = HTTPServer(DOCUMENT_ROOT, ADDRESS, PORT)
        self.thread = Thread(target=self.httpd.serve_forever)
        self.thread.start()
        self['URL'] = 'http://localhost:%i/' % PORT

    def tearDown(self):
        self.httpd.shutdown()
        self.httpd.server_close()
        while self.thread.isAlive():
            pass
        self.thread.join()


SERVER_LAYER = ServerLayer()


ftw/maintenanceserver/server.py
from SimpleHTTPServer import SimpleHTTPRequestHandler
import SocketServer
import os
import posixpath
import re
import socket
import urllib


def remove_virtual_host_monster_config(path):
    if 'VirtualHostBase' not in path.split('/'):
        return path

    path = re.sub(r'^\/VirtualHostBase\/.*\/VirtualHostRoot', r'', path)
    while path.startswith('/_vh_'):
        path = re.sub(r'/(_vh_[^/]+)/', '/', path)
    return path


class HTTPRequestHandler(SimpleHTTPRequestHandler):

    def __init__(self, request, client_address, server):
        self.request = request
        self.client_address = client_address
        self.server = server
        self.setup()

        # Do not finish() when we have a broken pipe.
        try:
            self.handle()
        except socket.error:
            pass  # broken pipe
        except:
            self.finish()
            raise
        else:
            self.finish()

    do_POST = SimpleHTTPRequestHandler.do_GET

    def do_OPTIONS(self):
        self.send_response(200, method='OPTIONS')
        self.send_header('Allow', 'GET,HEAD,POST,OPTIONS')
        self.send_header('Cache-Control', 'no-cache')
        self.end_headers()

    def send_response(self, code, message=None, method=None):
        # Send "503 Service Unavailable" instead of "200 OK" so that
        # caching proxies do not cache maintenance server responses,
        # except to resources request.
        if code == 200 and method != 'OPTIONS' and not self.is_resource:
            code = 503
            message = None

        return SimpleHTTPRequestHandler.send_response(self, code, message)

    def translate_path(self, path):
        path = remove_virtual_host_monster_config(path)
        # abandon query parameters
        path = path.split('?',1)[0]
        path = path.split('#',1)[0]
        path = posixpath.normpath(urllib.unquote(path))
        words = path.split('/')
        words = filter(None, words)

        document_root = os.path.abspath(self.server.document_root)
        index = os.path.join(document_root, 'index.html')
        filepath = os.path.abspath(os.path.join(document_root, *words))

        if not os.path.isfile(filepath):
            self.is_resource = False
            return index

        elif not filepath.startswith(document_root):
            self.is_resource = False
            return index

        else:
            self.is_resource = True
            return filepath


class HTTPServer(SocketServer.TCPServer):

    allow_reuse_address = True

    def __init__(self, document_root, address, port):
        self.document_root = document_root
        SocketServer.TCPServer.__init__(
            self, (address, port), HTTPRequestHandler)


def command(document_root, port, address='localhost'):
    HTTPServer(document_root, address, port).serve_forever()
